app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClientIntegration.java

```diff
@@ -0,0 +1,184 @@
+package org.wikipedia.analytics.eventplatform;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import org.wikipedia.WikipediaApp;
+import org.wikipedia.dataclient.ServiceFactory;
+import org.wikipedia.dataclient.WikiSite;
+import org.wikipedia.dataclient.okhttp.OkHttpConnectionFactory;
+import org.wikipedia.settings.Prefs;
+import org.wikipedia.util.log.L;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
+import io.reactivex.rxjava3.disposables.CompositeDisposable;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import retrofit2.Retrofit;
+import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+import static java.net.HttpURLConnection.HTTP_ACCEPTED;
+import static java.net.HttpURLConnection.HTTP_BAD_REQUEST;
+import static java.net.HttpURLConnection.HTTP_CREATED;
+import static java.net.HttpURLConnection.HTTP_GATEWAY_TIMEOUT;
+import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
+import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
+import static org.wikipedia.BuildConfig.META_WIKI_BASE_URI;
+import static org.wikipedia.json.GsonUtil.getDefaultGson;
+import static org.wikipedia.settings.Prefs.getEventPlatformIntakeUriOverride;
+import static org.wikipedia.settings.Prefs.getEventPlatformSessionId;
+import static org.wikipedia.settings.Prefs.getStreamConfigs;
+import static org.wikipedia.settings.Prefs.setEventPlatformSessionId;
+import static org.wikipedia.settings.Prefs.setStreamConfigs;
+import static org.wikipedia.util.DateUtil.iso8601DateFormat;
+
+final class EventPlatformClientIntegration {
+
+    private static final Map<String, EventService> RETROFIT_SERVICE_CACHE = new HashMap<>();
+    private static final CompositeDisposable DISPOSABLES = new CompositeDisposable();
+
+    static void fetchStreamConfigs(EventPlatformClient.StreamConfigsCallback cb) {
+        DISPOSABLES.add(ServiceFactory.get(new WikiSite(META_WIKI_BASE_URI))
+                .getStreamConfigs()
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(response -> cb.onSuccess(response.getStreamConfigs()), L::e));
+    }
+
+    /**
+     * Lazily instantiate a Retrofit service instance for the destination event service specified in
+     * the provided stream configuration.
+     *
+     * @param streamConfig stream configuration
+     * @return EventService
+     */
+    @NonNull
+    static EventService getEventService(@NonNull StreamConfig streamConfig) {
+        DestinationEventService dest = streamConfig.getDestinationEventService();
+        if (dest == null) {
+            dest = DestinationEventService.ANALYTICS;
+        }
+
+        EventService service;
+        if (RETROFIT_SERVICE_CACHE.containsKey(dest.getId())) {
```

Comment:
`Retrofit` service objects are very heavy and expensive. Is there a reason we need separate service objects for each and every different stream config?

Reply:
#1818 makes changes to this.